File: src/block/Block.java
package block;

public abstract class Block {
    public final BlockAttribute[] attributes;
    public final String name;
    public final int totalStates;
    private boolean initialized;

    protected Block(String name, BlockAttribute... attributes) {
        //noinspection ConstantConditions
        if (this.initialized) throw new RuntimeException("Cannot reconstruct class");
        this.initialized = true;

        this.name = name;
        this.attributes = attributes;
        int states = 1;
        for (BlockAttribute attribute : this.attributes) {
            states *= attribute.property().count + 1;
        }
        this.totalStates = states;
    }

    protected int getTotalStates() {
        return this.totalStates;
    }

    protected int genStates(BlockState[] registry, int index) {
        byte[][] states = new byte[this.totalStates][this.attributes.length];

        int blockSize = 1;
        for (int prop = 0; prop < this.attributes.length; prop++) {
            int size = this.attributes[prop].property().count + 1;
            for (int i = 0; i < this.totalStates; i++) {
                states[i][prop] = (byte) ((i / blockSize) % size);
            }
            blockSize *= size;
        }
        for (int i = 0; i < this.totalStates; index++) {
            registry[index] = new BlockState(this, states[i++]);
        }
        return index;
    }

    protected abstract boolean isFullCube(BlockState blockState);

    public abstract BlockModel getModel(BlockState blockState);
}

File: java/src/block/Block.kt
@file:Suppress("UNCHECKED_CAST")

package src.block

abstract class Block protected constructor(name: String, vararg attributes: BlockAttribute) {
    val attributes: Array<BlockAttribute>
    val name: String
    val totalStates: Int

    init {
        this.name = name
        this.attributes = attributes as Array<BlockAttribute>
        var states = 1
        for (attribute in this.attributes) {
            states *= attribute.property.count + 1
        }
        totalStates = states
    }

    fun genStates(registry: Array<BlockState>, start: Int): Int {
        var end = start
        val states = Array(totalStates) { ByteArray(attributes.size) }
        var blockSize = 1
        for (prop in attributes.indices) {
            val size = attributes[prop].property.count + 1
            for (i in 0 until totalStates) {
                states[i][prop] = (i / blockSize % size).toByte()
            }
            blockSize *= size
        }
        var i = 0
        while (i < totalStates) {
            registry[end++] = BlockState(this, *states[i++])
        }
        return end
    }

    abstract fun getModel(blockState: BlockState): BlockModel
    abstract fun hasModel(blockState: BlockState): Boolean
}